File: crates/ffi/lang/python/examples/example_project/build_plugin.py
"""
Build a Rust plugin (`.so`/`.dylib`/`.dll`) from Python nodes (no manifest output by default).

Run (from repo root):
  python crates/ffi/lang/python/examples/example_project/build_plugin.py /tmp/example_py.so
"""

from __future__ import annotations

import sys
from pathlib import Path

# Import the in-repo SDK directly.
REPO_ROOT = Path(__file__).resolve().parents[6]
PY_SDK = REPO_ROOT / "crates" / "ffi" / "lang" / "python"
sys.path.insert(0, PY_SDK.as_posix())

from daedalus_py import Plugin  # noqa: E402

from nodes import register_all  # noqa: E402


def main() -> None:
    out = Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else (Path.cwd() / "example_py.so")
    out.parent.mkdir(parents=True, exist_ok=True)

    plugin = Plugin(name="example_py", version="0.1.0", description="Python example project")
    register_all(plugin)

    # Point runtime at the nodes module.
    nodes_path = Path(__file__).with_name("nodes.py").resolve().as_posix()
    for n in plugin.nodes:
        n.py_function = n.id.split(":")[-1]
        n.py_path = nodes_path

    artifact = plugin.build(out_path=out, out_name="example_py", bundle=True, release=True)
    print(artifact.as_posix())


if __name__ == "__main__":
    main()

File: crates/ffi/lang/node/examples/example_project/js/build_plugin.mjs
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import process from "node:process";

// Import the in-repo SDK directly.
const repoRoot = path.resolve(path.dirname(new URL(import.meta.url).pathname), "../../../../../../..");
const sdk = path.join(repoRoot, "crates", "ffi", "lang", "node", "daedalus_node", "index.mjs");
const { Plugin, NodeDef, t } = await import(`file://${sdk}`);

const out = process.argv[2]
  ? path.resolve(process.argv[2])
  : path.join(os.tmpdir(), "example_node.so");
fs.mkdirSync(path.dirname(out), { recursive: true });

const plugin = new Plugin({ name: "example_node", version: "0.1.1", description: "Node example project" });

// Point to the local nodes module sitting next to this script.
const nodesPath = path.join(path.dirname(new URL(import.meta.url).pathname), "nodes.mjs");
plugin.runtime({ js_path: nodesPath });

plugin.register(
  new NodeDef({
    id: "example_node:add",
    js_function: "add",
    inputs: [
      { name: "a", ty: t.int() },
      { name: "b", ty: t.int() },
    ],
    outputs: [{ name: "out", ty: t.int() }],
    metadata: { lang: "node", kind: "stateless" },
  }),
);

plugin.register(
  new NodeDef({
    id: "example_node:split",
    js_function: "split",
    inputs: [{ name: "value", ty: t.int() }],
    outputs: [
      { name: "out0", ty: t.int() },
      { name: "out1", ty: t.int() },
    ],
    metadata: { lang: "node", kind: "multi_output" },
  }),
);

plugin.register(
  new NodeDef({
    id: "example_node:counter",
    js_function: "counter",
    inputs: [{ name: "inc", ty: t.int() }],
    outputs: [{ name: "out", ty: t.int() }],
    stateful: true,
    state: { start: 0 },
    metadata: { lang: "node", kind: "stateful" },
  }),
);

const artifact = plugin.build({ out_path: out, out_name: "example_node", bundle: true, bundle_deps: true, release: true });
process.stdout.write(artifact + "\n");
